### PR TITLE
Refactor promoteType to no longer use shifting strategy

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #116810
* #116809
* #116808
* #116807
* #116806
* #116805
* #116804
* #116803
* #116595
* #116594
* __->__ #116693
* #116698

Instead of manually fixing the indices (extremely error prone when new
dtypes are added) we just setup a lookup table to map ScalarType to the
offsets table.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

### CODE REVIEW

**Review by albanD — APPROVED** [review]
Nice cleamup!

**Comment by ezyang** [comment]
Apparently my checkout was really old and I had to rebase past @drisspg's refactor at https://github.com/pytorch/pytorch/pull/113975 but the idea here still applies.

**Review by albanD — APPROVED** [review]

File: c10/core/ScalarType.cpp [hunk 1]
```diff
@@ -1,23 +1,48 @@
...
+constexpr auto f4 = ScalarType::Float;
+constexpr auto f8 = ScalarType::Double;
+constexpr auto c2 = ScalarType::ComplexHalf;
+constexpr auto c4 = ScalarType::ComplexFloat;
+constexpr auto c8 = ScalarType::ComplexDouble;
+constexpr auto b1 = ScalarType::Bool;
+constexpr auto bf = ScalarType::BFloat16;
+constexpr auto ud = ScalarType::Undefined;
+
+constexpr int64_t NUM_PROMOTE_TYPES = 20;
```
Comment (Skylion007) [inline c10/core/ScalarType.cpp:23]: Couldn't you use array_of to construct the array below and then set this constant by exploiting the fact the std::array::size is a constexpr? Wouldn't this make it less error prone to forgetting to update this in the future potentially?

Comment (ezyang) [reply c10/core/ScalarType.cpp:23]: Oh I see, it's not a stdlib thing but there's a helper hanging out in an inconvenient place. I guess I'll move it over.

**Pull request closed without merging**